File: app/src/main/java/ru/melod1n/vk/service/LongPollService.java
package ru.melod1n.vk.service;

import android.app.Service;
import android.content.Intent;
import android.os.IBinder;
import android.util.ArrayMap;
import android.util.Log;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import org.json.JSONArray;
import org.json.JSONObject;

import ru.melod1n.vk.api.VKApi;
import ru.melod1n.vk.api.model.VKLongPollServer;
import ru.melod1n.vk.api.LongPollParser;
import ru.melod1n.vk.concurrent.LowThread;
import ru.melod1n.vk.net.HttpRequest;
import ru.melod1n.vk.util.Util;

public class LongPollService extends Service {

    private static final String TAG = "LongPollService";

    private Thread thread;
    private boolean running;

    @Override
    public void onCreate() {
        super.onCreate();

        running = false;
        thread = new LowThread(new Updater());
    }

    @Nullable
    @Override
    public IBinder onBind(Intent intent) {
        return null;
    }

    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        if ((flags & START_FLAG_RETRY) == 0) {
            Log.w(TAG, "Retry launch!");
        } else {
            Log.d(TAG, "Simple launch");
        }

        if (running) return START_STICKY;

        running = true;

        try {
            thread.start();
        } catch (Exception e) {
            e.printStackTrace();
        }

        return START_STICKY;
    }

    @Override
    public void onDestroy() {
        super.onDestroy();

        running = false;
        thread.interrupt();
    }

    private class Updater implements Runnable {
        @Override
        public void run() {
            VKLongPollServer server = null;
            while (running) {
                if (!Util.hasConnection()) {

                    try {
                        Thread.sleep(5_000);
                    } catch (InterruptedException e) {
                        e.printStackTrace();
                    }
                    continue;
                }

                try {
                    if (server == null) {
                        server = VKApi.messages().getLongPollServer().execute(VKLongPollServer.class).get(0);
                    }

                    JSONObject response = getResponse(server);
                    if (response == null || response.has("failed")) {
                        Log.w(TAG, "Failed get response from");
                        Thread.sleep(1_000);
                        server = null;
                        continue;
                    }

                    long tsResponse = response.optLong("ts");
                    JSONArray updates = response.getJSONArray("updates");

                    Log.i(TAG, "updates: " + updates);

                    server.ts = tsResponse;
                    if (updates.length() != 0) {
                        process(updates);
                    }
                } catch (Exception e) {
                    e.printStackTrace();
                    try {
                        Thread.sleep(5_000);
                        server = null;
                    } catch (InterruptedException e1) {
                        e1.printStackTrace();
                    }
                }

            }
        }

        private JSONObject getResponse(@NonNull VKLongPollServer server) throws Exception {
            ArrayMap<String, String> params = new ArrayMap<>();
            params.put("act", "a_check");
            params.put("key", server.key);
            params.put("ts", String.valueOf(server.ts));
            params.put("wait", "10");
            params.put("mode", "490");
            params.put("version", "9");

            String buffer = HttpRequest.get("https://" + server.server, params).asString();
            return buffer == null ? null : new JSONObject(buffer);
        }


        private void process(JSONArray updates) {
            LongPollParser.getInstance().parse(updates);
        }

    }
}

File: app/src/main/java/ru/melod1n/vk/service/LongPollService.kt
package ru.melod1n.vk.service

import android.app.Service
import android.content.Intent
import android.os.IBinder
import android.util.ArrayMap
import android.util.Log
import org.json.JSONArray
import org.json.JSONObject
import ru.melod1n.vk.api.UserConfig
import ru.melod1n.vk.api.VKApi.messages
import ru.melod1n.vk.api.VKLongPollParser
import ru.melod1n.vk.api.model.VKLongPollServer
import ru.melod1n.vk.concurrent.LowThread
import ru.melod1n.vk.net.HttpRequest
import ru.melod1n.vk.util.AndroidUtils

class LongPollService : Service() {
    private var thread: Thread? = null
    private var running = false
    override fun onCreate() {
        super.onCreate()
        running = false
        thread = LowThread(Updater())
    }

    override fun onBind(intent: Intent?): IBinder? {
        return null
    }

    override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
        if (flags and START_FLAG_RETRY == 0) {
            Log.w(TAG, "Retry launch!")
        } else {
            Log.d(TAG, "Simple launch")
        }
        if (running) return START_STICKY
        running = true
        try {
            thread!!.start()
        } catch (e: Exception) {
            e.printStackTrace()
        }
        return START_STICKY
    }

    override fun onDestroy() {
        super.onDestroy()
        running = false
        thread!!.interrupt()
    }

    private inner class Updater : Runnable {
        override fun run() {

            var server: VKLongPollServer? = null

            while (running && UserConfig.isLoggedIn) {
                if (!AndroidUtils.hasConnection()) {
                    try {
                        Thread.sleep(5000)
                    } catch (e: InterruptedException) {
                        e.printStackTrace()
                    }
                    continue
                }
                try {
                    if (server == null) {
                        server = messages().longPollServer.execute(VKLongPollServer::class.java)!![0]
                    }
                    val response = getResponse(server)
                    if (response == null || response.has("failed")) {
                        Log.w(TAG, "Failed get response from")
                        Thread.sleep(1000)
                        server = null
                        continue
                    }
                    val tsResponse = response.optLong("ts")
                    val updates = response.getJSONArray("updates")
                    Log.i(TAG, "updates: $updates")
                    server.ts = tsResponse
                    if (updates.length() != 0) {
                        process(updates)
                    }
                } catch (e: Exception) {
                    e.printStackTrace()
                    try {
                        Thread.sleep(5000)
                        server = null
                    } catch (e1: InterruptedException) {
                        e1.printStackTrace()
                    }
                }
            }
        }

        @Throws(Exception::class)
        private fun getResponse(server: VKLongPollServer): JSONObject? {
            val params = ArrayMap<String, String>()
            params["act"] = "a_check"
            params["key"] = server.key
            params["ts"] = server.ts.toString()
            params["wait"] = "10"
            params["mode"] = "490"
            params["version"] = "9"
            val buffer = HttpRequest["https://" + server.server, params].asString()
            return JSONObject(buffer)
        }

        private fun process(updates: JSONArray) {
            VKLongPollParser.parse(updates)
        }
    }

    companion object {
        private const val TAG = "LongPollService"
    }
}